src/lib/utils/object-graph.ts

```diff
@@ -0,0 +1,172 @@
+import { v4 as uuidv4 } from 'uuid'
+
+/**
+ * Represents a serialized object structure where circular/repeated references
+ * are replaced with `{ __refId }` and the full objects are provided in the
+ * `include` map.
+ */
+export interface FlattenedGraph<T> {
+  data: T
+  include: Record<string, unknown>
+}
+
+/**
+ * A reference to an object stored in the `include` section of a
+ * `DereferencedData` structure.
+ */
+export interface Reference {
+  __refId: string
+}
+
+/**
+ * Serializes an object containing circular and/or repeated references into a
+ * JSON-safe format by extracting duplicates into an `include` map.
+ */
+export function flattenObjectGraph<T>(input: T): FlattenedGraph<T> {
+  // Map of all seen objects. Each object gets a UUID and a flag for whether
+  // it's reused.
+  const seen = new Map<unknown, { __refId: string; isDuplicate: boolean }>()
+
+  // The final store of extracted objects that will be referenced by `{ __refId }`
+  const include: FlattenedGraph<T>['include'] = {}
+
+  // Guards against infinite recursion by tracking active __refIds during
+  // serialization
+  const serializing = new Set<string>()
+
+  /**
+   * First pass: detects duplicates and assigns UUIDs to all objects. Only marks
+   * objects as `isDuplicate` if seen more than once.
+   */
+  function findDuplicates(value: unknown) {
+    if (typeof value !== 'object' || value === null) return
+
+    const existing = seen.get(value)
+    if (existing) {
+      existing.isDuplicate = true
+      return
+    } else {
+      seen.set(value, { __refId: uuidv4(), isDuplicate: false })
+    }
+
+    if (Array.isArray(value)) {
+      for (const item of value) findDuplicates(item)
+    } else {
+      for (const val of Object.values(value)) {
+        findDuplicates(val)
+      }
+    }
+  }
+
+  /**
+   * Second pass: recursively walks the data and replaces duplicate objects
+   * with `{ __refId }`. If an object is marked as a duplicate, it is added to
+   * `include` and referenced by ID.
+   */
+  function walk(value: unknown): unknown {
+    if (typeof value !== 'object' || value === null) return value
+
+    const entry = seen.get(value)
+    if (entry?.isDuplicate) {
+      if (!(entry.__refId in include)) {
+        // If we're already serializing this __refId (due to circularity), just
+        // return the ref
```

Comment:
I didn't understand the order of events here when I read this comment and didn't realize what was happening until I saw the `remove` at the end. It might help me understand better if it said "If we're _currently_ serializing..."